File: Algorithm/ORM_Matlab/algos/sense_danger.m
function dangerObs = sense_danger(senseObs, robotParams, disDanger)
%Detect dangerous obstacle which can collide with robot
radRobot = robotParams(1);
[numObs, ~] = size(senseObs);
dangerObs = zeros(numObs, 4);
countDanger = 0;

for i=1:numObs
    disObs = senseObs(i, 1);
    radObs = senseObs(i, 3);
    dirObs = senseObs(i, 2);
    vObs = senseObs(i,4);
    disSafe = radObs + radRobot;
    disObs_real = disObs - disSafe;
    if  disObs_real < disDanger
        countDanger = countDanger + 1;
        try
            beta_high = abs(atan2(disSafe, sqrt(disObs^2 - disSafe^2)));
            beta_low = beta_high;
            if vObs ~= 0
                k1 = min(1, 150/(50+disObs_real));
                k2 = 1;
                del_angle = limit_angle(senseObs(i,5) - dirObs);              
                if del_angle > 0
                    
                    beta_high = pi/2;
%                     beta_high = min(beta_high*(k1 + k2*abs(cos(del_angle))), pi/2);
                    beta_low = min(beta_low*(k1), pi/2);
                elseif del_angle < 0
                    beta_high = min(beta_high*(k1), pi/2);
%                     beta_low = pi/2;
                    beta_low = min(beta_low*(k1 + k2*abs(cos(del_angle))), pi/2);
                end
                vObs = cos(del_angle)*vObs;
            end
        catch
            beta_high = pi/2;
            beta_low = beta_high;
        end
        rangeObs =  sort_range([limit_angle(dirObs+beta_high) , limit_angle(dirObs-beta_low)]);
        dangerObs(countDanger, :) = [disObs_real, rangeObs, vObs];
    end
end

if countDanger == 0
    dangerObs = [];
else
    dangerObs = dangerObs(1:countDanger,:);
end

function pair = sort_range(pair)
    if pair(1) > pair(2)
        pair = [pair(2), pair(1)];
    end
end

end
        
        

File: Algorithm/ORM_Python/algorithm/sense_danger.py
import numpy as np
from helper.calc_add_limit_vector import limit_angle


def sort_range(pair):
    if pair[0] > pair[1]:
        pair = [pair[1], pair[0]]
    return pair

""" Detect dangerous obstacle which can collide with robot """
def sense_danger(senseObs, robotParams, disDanger): # 23x5 , 1x5, 0
    radRobot = robotParams[0]
    numObs = senseObs.shape[0]
    dangerObs = np.zeros((numObs, 4), dtype=np.float64) 
    countDanger = 0
    
    for i in range(0, numObs):
        disObs = senseObs[i, 0]
        radObs = senseObs[i, 2]
        dirObs = senseObs[i, 1]
        vObs = senseObs[i,3]
        disSafe = radObs + radRobot
        disObs_real = disObs - disSafe
        if  disObs_real < disDanger:
            countDanger = countDanger + 1
            try:
                beta_high = abs(np.arctan2(disSafe, np.sqrt(np.power(disObs,2) - np.power(disSafe,2))))
                beta_low = beta_high
                if vObs != 0:
                    k1 = min(1, 150.0/(50.0+disObs_real))
                    k2 = 1
                    del_angle = limit_angle(senseObs[i,4] - dirObs)              
                    if del_angle > 0:
                        beta_high = np.pi/2
                        #beta_high = min(beta_high*(k1 + k2*abs(np.cos(del_angle))), np.pi/2)
                        beta_low = min(beta_low*k1, np.pi/2)
                    elif del_angle < 0:
                        beta_high = min(beta_high*k1, np.pi/2)
                        #beta_low = np.pi/2
                        beta_low = min(beta_low*(k1 + k2*abs(np.cos(del_angle))), np.pi/2)
                    vObs = np.cos(del_angle)*vObs
            except:
                beta_high = np.pi/2
                beta_low = beta_high  
            rangeObs =  sort_range([limit_angle(dirObs+beta_high) , limit_angle(dirObs-beta_low)])
            dangerObs[countDanger-1, :] = [disObs_real, rangeObs[0], rangeObs[1], vObs]
    
    if countDanger == 0:
        dangerObs = np.array([])  # np.zeros((10,4))
    else:
        dangerObs = dangerObs[0:countDanger,:]
    return dangerObs
            